src/app/CommandHandler.cpp

```diff
@@ -124,6 +123,10 @@ void CommandHandler::Close()
 {
     mSuppressResponse = false;
     MoveToState(CommandState::AwaitingDestruction);
+    // We must finished all works before we can shutdown a CommandHandler, the actual CommandHandler MUST finish their work in
+    // reasonable time or there is a bug.
```

Comment:
But more generally, I don't see how this can be a reasonable thing to VerifyOrDie on.  If the callee is in the middle of async processing a command and needs to reboot or power down or otherwise do a clean shutdown what can it do?  There is no facility to drop the ref without sending the reply (which it might not be able to do at this point) and there is no way to shut down cleanly with this VerifyOrDie....

It feels like we need to figure out how this bit is going to work.

Reply:
https://github.com/project-chip/connectedhomeip/pull/8197#discussion_r667186882

Per previous discussions, seems VerifyOrDie is suitable for this case.

Why do we need a clean shutdown when
- we want to reboot
- And we want to interrupt a existing transaction

Killing the process should be acceptable in this case?

Reply:
It depends on what else is in the process address space, no?

Matter might not be the only thing running in the process, and we might need to restart the Matter stack but don't want the whole process to go down.  Or maybe we are even rebooting, but we need to do other cleanup (e.g. persisting data) during that reboot.

Point is, a component like the SDK does not own the process and should not be killing it if it can be avoided.

Reply:
One example of what we could do instead, by the way (though I am not sure whether it's better): Handles form a doubly-linked list.  Destructor removes from the list if in the list.  Shutting down the stack goes through the list and nulls out the pointer in all the handles, and unlinks the list.  Consumers null-check the handle after going async to make sure it's not nulled out.

The big question is whether the resource consumption of that is OK.

Reply:
Instead, I would suggest implement the real graceful shutdown logic, which will block new commands, and wait for a reasonable time for existing command to finish.

Also, it is possible to make the application register a OnShutdown() callback, and release the resources they own before shut down?

Reply:
The problem is that the async refs might be held within the stack/SDK itself, not in the application, for some clusters.

Reply:
What if we regard the cluster implementation as application code?

IMHO just nullout the reference is not enough, if we may shutdown the chip stack while keep other things running in the same process.

As most async operations are waiting for a external callback, then what if we shutdown the chip stack and then the callback fired?

Thus I think a better solution is to notice the potential users of the aync response that "the CHIP stack is shutting down, you need to abort your external async work"

Reply:
> What if we regard all the cluster implementation as application code?

But the application doesn't necessarily know the cluster is doing async things and may not have access to ways to shut those things down.

> Could you provide some example for passing the async refs to the SDK?

How do we expect the EnableNetwork command to work?

> then what if we shutdown the chip stack and then the callback fired?

Then the code that was going to use the reference when it fires will null-check, discover it no longer has a valid handle, and do whatever it feels it should do in that case?

> Thus I think a better solution is to notice the potential users of the aync response 

How will you find them?

Reply:
> But the application doesn't necessarily know the cluster is doing async things and may not have access to ways to shut those things down.

The code that setup the async work should be able to shutdown the async work, at least mark it invalid and ignore the callback by itself.

> How do we expect the EnableNetwork command to work?

The EnableNetwork will should call some code under `src/platform`, and waiting for some event callback. But the command handler should be held by the NetworkCommissioningCluster instead of (passing to) other sdk code.

> Then the code that was going to use the reference when it fires will null-check, discover it no longer has a valid handle, and do whatever it feels it should do in that case?

I think that callback should never be called, since
1. If it comes from CHIP Stack, since we have stopped the event loop, it won't be called.
2. If it comes from some external code, it may holding some dangling references.

> How will you find them?

I would suggest add a API to CHIPStack for register some callbacks when shutdown is called, the cluster can register it when starting up. Or at least we can generate `ShutdownXxxxClusterCallback()`

Reply:
And if we still want to continue on

> One example of what we could do instead, by the way (though I am not sure whether it's better): Handles form a doubly-linked list. Destructor removes from the list if in the list. Shutting down the stack goes through the list and nulls out the pointer in all the handles, and unlinks the list. Consumers null-check the handle after going async to make sure it's not nulled out.

I'm thinking about adding a magic number to the IM Engine (or CHIPStack) to mark the current "boot". Then everytimg we dereference the Handle, we check if the magic number matches the current magic number on IMEngine (i.e. a lazy weak pointer).

This approach should be straight forward, and safe for now.

Reply:
Updated the PR with a magic number approach, this should work as long as stack shutting down is the only case when the underlying object is invalid for the object handle and hope this will be correct.

Regarding we already have the init function for some cluster handlers, I believe we need a proper cluster shutdown oneday.

Reply:
> The code that setup the async work should be able to shutdown the async work

The code that set up the async work may not be the code that triggers stack shutdown. They may not really know about each other at all.

> But the command handler should be held by the NetworkCommissioningCluster

Indeed, and the point is _the app does not have a reference to it_.

> If it comes from some external code, it may holding some dangling references.

Yes, the point is to make sure the external code does not hold dangling references.

I'll take a look at the updated code tomorrow....